backend/uasmartsignage/src/main/java/deti/uas/uasmartsignage/Services/TemplateGroupService.java

```diff
@@ -130,6 +157,92 @@ public TemplateGroup updateTemplateGroup(Long id, TemplateGroup templateGroup) {
         return templateGroupRepository.save(templateGroupById);
     }
 
-    
-    
+    /**
+     * Takes a template and the contents to place in the widgets and returns the full html
+     * 
+     * @param template The Template containing the list of widgets
+     * @param contents A Map that contains the ids of Contents and the respective variables to fill in the widgets
+     * @param monitorWidth The screen width to be considered when calculating widget sizes
+     * @param monitorHeight The screen height to be considered when calculating widget sizes
+     * @return The created HTML in a String, or {@code null} if creation fails.
+     */
+    public String generateHTML(Template template, Map<Integer, String> contents, int monitorWidth, int monitorHeight) {
+
+        List<TemplateWidget> widgets = template.getTemplateWidgets();
+        String filePath = "static/base.html";
+
+        try {
+            
+            // get a base html to add the widgets to
+            File baseFile = ResourceUtils.getFile("classpath:" + filePath);
+            Document doc = Jsoup.parse(baseFile, "UTF-8");
+
+            for (TemplateWidget widget : widgets) {
+                // fill in the variables inside the widgets
+                // using the values inside "contents"
+                String widgetHTML = loadWidget(widget, contents, monitorWidth, monitorHeight);
+                
+                if(widgetHTML == null){
+                    continue;
+                }   
+
+                // add the html of the widget to the main body
+                List<Element> widgetElements = Jsoup.parseBodyFragment(widgetHTML).body().children();
+                for(Element child : widgetElements)
+                    doc.body().appendChild(child);
+            }
+
+            return doc.html();
```

Comment:
Maybe adding a log to indicate that it was successfully generated could be useful